task/sensor-api/src/lib.rs

```diff
@@ -8,26 +8,107 @@
 
 use userlib::*;
 
-pub use task_sensor_types::*;
+use core::convert::TryFrom;
+
+use hubpack::SerializedSize;
+use serde::{Deserialize, Serialize};
+pub use task_sensor_types::{NoData, Reading, SensorError};
+
+#[derive(
+    zerocopy::AsBytes,
+    Copy,
+    Clone,
+    Debug,
+    Eq,
+    PartialEq,
+    Serialize,
+    SerializedSize,
+)]
+#[repr(C)]
+pub struct SensorId(u32);
+
+pub struct InvalidSensor;
+
+impl SensorId {
+    /// The maximum allowable `SensorId` value on this system.
+    pub const MAX_VALUE: u32 = config::NUM_SENSORS as u32;
+
+    /// Returns a new `SensorId` for the provided `u32` value.
+    ///
+    /// # Panics
+    ///
+    /// If `id` is greater than [`config::NUM_SENSORS`].
+    #[must_use]
+    pub const fn new(id: u32) -> Self {
+        // On devboard targets without sensors, `NUM_SENSORS` is zero,
```

Comment:
Nit: now that we have `TryFrom<u32>`, should we replace this with `id.try_into().unwrap()` (or `unwrap_lite()`)?

Reply:
This is a `const fn`, so we can't use `try_into()`, as it's a non-`const` trait method.. Potentially, we could add a non-trait method `const fn try_new(id: u32) -> Result<Self, InvalidSensorId>`, and call that (and have the `TryFrom` impl call it as well) but we would need to `match` on its result rather than `unwrap()` (which isn't `const` as it formats the error) or `unwrap_lite() ( which isn't `const` as it's a trait method). I'll go and do that, since I agree that it's nice to not have two copies of the validation.